src/runtime_src/xdp/profile/plugin/aie_debug/aie_debug_plugin.cpp

```diff
@@ -0,0 +1,401 @@
+          uint32_t coreStatus = 0;
+          auto tileOffset = _XAie_GetTileAddr(aieDevInst, tile.row + 1, tile.col);
+          XAie_Read32(aieDevInst, tileOffset + AIE_OFFSET_CORE_STATUS, &coreStatus);
+
+          auto& coreStallCounter = coreStuckCountMap[tile];
+
+          // Condition : Core is in reset/done state or not enabled
+          if (coreStatus & CORE_INACTIVE_MASK || !(coreStatus & CORE_ENABLE_MASK)) {
+            coreUnstalled = (coreStallCounter >= GRAPH_HANG_COUNT_THRESHOLD);
+            coreStallCounter = 0;
+          }
+          // Condition : If core is enabled + stalled and has same kind of stall as previous check
+          else if ((coreStatus & CORE_STALL_MASK) && (coreStatus == coreStatusMap[tile]) ) {
+            coreStallCounter++;
+          }
+          // Core is running normally or has changed state
+          else {
+            coreUnstalled = (coreStallCounter >= GRAPH_HANG_COUNT_THRESHOLD);
+            coreStallCounter = 0;
+          }
+
+          // Is this core contributing to entire graph hang?
+          if (coreUnstalled && graphStallCounter) {
+            graphStallCounter--;
+          } else if (coreStallCounter == GRAPH_HANG_COUNT_THRESHOLD) {
+            graphStallCounter++;
+          }
+
+          // Is this core stuck for long time?
+          if (coreStallCounter == CORE_HANG_COUNT_THRESHOLD) {
+            foundStuckCores = true;
+            stuckTile = tile;
+            stuckCoreStatus = coreStatus;
+          }
+
+          coreStatusMap[tile] = coreStatus;
+        } // For tiles in graph
+
+        std::stringstream warningMessage;
+        if (graphStallCounter == graphTilesVec.size()) {
+          // We have a stuck graph
+          warningMessage
+          << "Potential deadlock/hang found in AI Engines. Graph : " << graphName;
+
+          xrt_core::message::send(severity_level::warning, "XRT", warningMessage.str());
+          // Send next warning if all tiles come out of hang & reach threshold again
+          graphStallCounter = 0;
+        } else if (foundStuckCores) {
+          // We have a stuck core within this graph
+          warningMessage
+          << "Potential deadlock/hang found in AI Engines. Graph : " << graphName << " "
+          << "Tile : " << "(" << stuckTile.col << "," << stuckTile.row + 1 << ") "
+          << "Status 0x" << std::hex << stuckCoreStatus << std::dec
+          << " : " << getCoreStatusString(stuckCoreStatus);
+
+          xrt_core::message::send(severity_level::warning, "XRT", warningMessage.str());
+          foundStuckCores = false;
+        }
+
+        // Print status for debug
+        if (xrt_core::config::get_verbosity() >= static_cast<unsigned int>(severity_level::debug)) {
+          std::stringstream msg;
+          for (const auto& tile : graphTilesVec) {
+            if (coreStuckCountMap[tile]) {
+              msg
+                << "T(" << tile.col <<"," << tile.row + 1 << "):" << "<" << coreStuckCountMap[tile]
```

Comment:
You might wish to use the boost format to format your output.
```suggestion
                << boost::format("T(%d,%d):<%d:0x%x>") % tile.col % tile.row + 1  % coreStuckCountMap[tile] % coreStatusMap[tile];
```